ffi/src/lib.rs

```diff
@@ -709,6 +757,334 @@ pub extern "C" fn visit_expression_literal_long(
+    let boxed_scan = unsafe { Box::from_raw(scan) };
+    let scan = boxed_scan.kernel_scan;
+    let scan_data = scan.scan_data(engine_interface.table_client().as_ref())?;
+    let data = KernelScanDataIterator {
+        data: Box::new(scan_data),
+        engine_interface,
+    };
+    Ok(data.into_handle())
+}
+
+/// # Safety
+///
+/// The iterator must be valid (returned by [kernel_scan_data_init]) and not yet freed by
+/// [kernel_scan_data_free]. The visitor function pointer must be non-null.
+#[no_mangle]
+pub unsafe extern "C" fn kernel_scan_data_next(
+    data: &mut KernelScanDataIterator,
+    engine_context: *mut c_void,
+    engine_visitor: extern "C" fn(
+        engine_context: *mut c_void,
+        engine_data: *mut EngineDataHandle,
+        selection_vector: KernelBoolSlice,
+    ),
+) -> ExternResult<bool> {
+    kernel_scan_data_next_impl(data, engine_context, engine_visitor)
+        .into_extern_result(data.engine_interface.error_allocator())
+}
+fn kernel_scan_data_next_impl(
+    data: &mut KernelScanDataIterator,
+    engine_context: *mut c_void,
+    engine_visitor: extern "C" fn(
+        engine_context: *mut c_void,
+        engine_data: *mut EngineDataHandle,
+        selection_vector: KernelBoolSlice,
+    ),
+) -> DeltaResult<bool> {
+    if let Some((data, sel_vec)) = data.data.next().transpose()? {
+        let bool_slice: KernelBoolSlice = sel_vec.into();
+        let data_handle = BoxHandle::into_handle(EngineDataHandle { data });
+        (engine_visitor)(engine_context, data_handle, bool_slice);
+        // ensure we free the data
+        unsafe { BoxHandle::drop_handle(data_handle) };
+        Ok(true)
+    } else {
+        Ok(false)
+    }
+}
+
+/// # Safety
+///
+/// Caller is responsible for (at most once) passing a valid pointer returned by a call to
+/// [kernel_scan_files_init].
+// we should probably be consistent with drop vs. free on engine side (probably the latter is more
+// intuitive to non-rust code)
+#[no_mangle]
+pub unsafe extern "C" fn kernel_scan_data_free(data: *mut KernelScanDataIterator) {
+    BoxHandle::drop_handle(data);
+}
+
+type CScanCallback = extern "C" fn(
+    engine_context: *mut c_void,
+    path: KernelStringSlice,
+    size: i64,
+    dv_info: *mut CDvInfo,
+    partition_map: *mut CStringMap,
+);
+
+pub struct CDvInfo {
+    dv_info: DvInfo,
+}
+impl BoxHandle for CDvInfo {}
+
+pub struct CStringMap {
+    values: HashMap<String, String>,
+}
+impl BoxHandle for CStringMap {}
+
+#[no_mangle]
+/// allow probing into a CStringMap. If the specified key is in the map, kernel will call
+/// allocate_fn with the value associated with the key and return the value returned from that
+/// function. If the key is not in the map, this will return NULL
+///
+/// # Safety
+///
+/// The engine is responsible for providing a valid [`CStringMap`] pointer and [`KernelStringSlice`]
+pub unsafe extern "C" fn get_from_map(
```

Comment:
In my testing with the dataset from `delta-kernel-rs/kernel/tests/data/basic_partitioned`, I'm not seeing the `__HIVE_DEFAULT_PARTITION__` partition passed through the kernel here. Perhaps this is handled in a special way?

Reply:
Hrmm, that could be a bug in our partition handling. I need to look at how that should be handled, thanks!

Reply:
Ahh. So afaict, `__HIVE_DEFAULT_PARTITION__` will only appear in path names, and that just happens when the partition value is `null`.

So you might see an action like:
```json
{"add":{"path":"letter=__HIVE_DEFAULT_PARTITION__/date=__HIVE_DEFAULT_PARTITION__/data=__HIVE_DEFAULT_PARTITION__/part-00000-701d9c03-c70a-4440-9e63-9b946d42fb69.c000.snappy.parquet","partitionValues":{"letter":null,"date":null,"data":null},"size":490,"modificationTime":1711048863762,"dataChange":true,"stats":"{\"numRecords\":1,\"minValues\":{\"number\":5},\"maxValues\":{\"number\":5},\"nullCount\":{\"number\":0}}"}}
```

So you just need to read at that path, and fill in `null` values for those columns. i.e. if you're selecting `"letter"`, then probing the `CStringMap` for `"letter"` will return `null`, so you can fill in that column with `null`